File: packages/spooky_core/lib/src/query/membership.dart
import '../state/lifecycle.dart' show QueryPhase;
import '../types.dart';

/// What a membership read may do with the id-set it got back.
enum MembershipOutcome {
  /// The set is the server's answer; commit it.
  applied,

  /// The server's `_00_query` row is gone while we hold rows.
  viewLost,

  /// The read proves nothing; keep what we have.
  ignored,
}

/// The durable `_00_view` row: the last server-confirmed membership.
class DurableView {
  const DurableView({required this.ids, required this.confirmed});
  final RecordVersionArray ids;
  final bool confirmed;
}

DurableView? parseViewRow(Map<String, dynamic>? row) {
  if (row == null) return null;
  final ids = row['ids'];
  if (ids is! List) return null;
  final pairs = <RecordVersion>[];
  for (final entry in ids) {
    if (entry is List && entry.length >= 2) {
      pairs.add((entry[0].toString(), (entry[1] as num).toInt()));
    }
  }
  return DurableView(ids: pairs, confirmed: row['confirmed'] == true);
}

/// "Resolved before": the server answered this query on this device at some
/// point. An unconfirmed empty row cannot be told apart from one written before
/// the marker existed, so it does not count.
bool isResolvedBefore(DurableView? view) =>
    view != null && (view.ids.isNotEmpty || view.confirmed);

ServerViewMeta metaFromRow(Object? row) {
  if (row is! Map) return ServerViewMeta.absent;
  final rawState = row['state'];
  final state = rawState == 'materializing' || rawState == 'ready'
      ? rawState as String
      : null;
  final rowCount = row['rowCount'];
  return ServerViewMeta(
    present: true,
    rowCount: rowCount is num ? rowCount.toInt() : null,
    state: state,
  );
}

/// Collapse duplicate `(id, version)` pairs to one per id, keeping the highest
/// version.
RecordVersionArray dedupeRecordVersions(RecordVersionArray pairs) {
  if (pairs.length < 2) return pairs;
  final best = <String, int>{};
  for (final (id, version) in pairs) {
    final prev = best[id];
    if (prev == null || version > prev) best[id] = version;
  }
  if (best.length == pairs.length) return pairs;
  return [for (final e in best.entries) (e.key, e.value)];
}

class MembershipDecisionInput {
  const MembershipDecisionInput({
    required this.phase,
    required this.held,
    required this.remoteArray,
    this.meta,
    this.verifiedRemoval = false,
  });

  final QueryPhase phase;

  /// Number of ids currently held as membership.
  final int held;
  final RecordVersionArray remoteArray;
  final ServerViewMeta? meta;
  final bool verifiedRemoval;
}

/// Whether a server id-set may be applied. A NON-EMPTY set is always the
/// server's answer. An EMPTY set is believed only when the server stands behind
/// it: the `_00_query` row is present, `ready` (or pre-`state`), and reports
/// zero rows. A missing row while we hold membership is a lost view.
MembershipOutcome decideMembershipOutcome(MembershipDecisionInput input) {
  if (input.remoteArray.isNotEmpty || input.verifiedRemoval) {
    return MembershipOutcome.applied;
  }
  final meta = input.meta;
  if (meta == null || !meta.present) {
    return input.held > 0 || input.phase != QueryPhase.cold
        ? MembershipOutcome.viewLost
        : MembershipOutcome.ignored;
  }
  final knownEmpty = meta.rowCount == 0 && meta.state != 'materializing';
  return knownEmpty ? MembershipOutcome.applied : MembershipOutcome.ignored;
}

// ---- list_ref snapshots ------------------------------------------------------

class ListRefSnapshot {
  ListRefSnapshot({
    required this.primary,
    required this.subquery,
    required this.meta,
  });

  RecordVersionArray primary;
  RecordVersionArray subquery;
  ServerViewMeta meta;
}

RecordVersionArray _toPairs(Object? rows) {
  if (rows is! List) return const [];
  final pairs = <RecordVersion>[
    for (final r in rows)
      if (r is Map)
        (r['out'].toString(), ((r['version'] as num?) ?? 0).toInt()),
  ];
  return dedupeRecordVersions(pairs);
}

/// Fold the single-query statement batch (edges, meta, children) into a
/// snapshot. Returns null when the edge statement did not answer with a list,
/// which the caller treats as a failed read rather than an empty one.
ListRefSnapshot? snapshotFromSingle(
    Object? items, Object? metaRow, Object? children) {
  if (items is! List) return null;
  return ListRefSnapshot(
    primary: _toPairs(items),
    subquery: _toPairs(children),
    meta: metaFromRow(metaRow),
  );
}

/// Fold the many-query statement batch into one snapshot per hash. Every hash
/// in [hashById] gets a snapshot; a query whose row did not come back reads
/// `present: false`.
Map<String, ListRefSnapshot> snapshotsFromBatch(
  Object? edges,
  Object? counts,
  Map<String, String> hashById,
) {
  final out = <String, ListRefSnapshot>{};
  if (edges is! List) return out;
  for (final hash in hashById.values) {
    out[hash] = ListRefSnapshot(
      primary: [],
      subquery: [],
      meta: ServerViewMeta.absent,
    );
  }
  for (final row in edges) {
    if (row is! Map) continue;
    final hash = hashById[row['in'].toString()];
    if (hash == null) continue;
    final snapshot = out[hash]!;
    final pair =
        (row['out'].toString(), ((row['version'] as num?) ?? 0).toInt());
    if (row['parent'] == null) {
      snapshot.primary.add(pair);
    } else {
      snapshot.subquery.add(pair);
    }
  }
  for (final snapshot in out.values) {
    snapshot.primary = dedupeRecordVersions(snapshot.primary);
    snapshot.subquery = dedupeRecordVersions(snapshot.subquery);
  }
  if (counts is List) {
    for (final count in counts) {
      if (count is! Map || count['id'] == null) continue;
      final hash = hashById[count['id'].toString()];
      if (hash == null) continue;
      out[hash]!.meta = metaFromRow(count);
    }
  }
  return out;
}

/// A batch that must not be believed at face value: a query we hold rows for
/// whose `_00_query` row did not come back, or no edge at all while the server
/// still reports rows for a held query. Such hashes are re-read one at a time.
List<String> suspectHashes(
  Map<String, ListRefSnapshot> snapshots,
  Map<String, int> heldCounts,
  int edgeCount,
) {
  final out = <String>[];
  for (final entry in snapshots.entries) {
    final held = heldCounts[entry.key] ?? 0;
    if (held == 0) continue;
    final meta = entry.value.meta;
    if (!meta.present || (edgeCount == 0 && (meta.rowCount ?? 0) > 0)) {
      out.add(entry.key);
    }
  }
  return out;
}

/// The record ids a stored `_00_view` / `_00_window` row names. Used by the
/// orphan collector, which only needs the ids and not their versions.
Iterable<String> decodeIdsOfView(Map<String, dynamic> row) sync* {
  final ids = row['ids'];
  if (ids is! List) return;
  for (final entry in ids) {
    if (entry is List && entry.isNotEmpty) yield entry.first.toString();
  }
}

File: packages/core/src/query/membership.ts
import type { RecordId } from 'surrealdb';
import type { MembershipOutcome, RecordVersionArray, ServerViewMeta } from '../types';
import type { QueryPhase } from '../state/lifecycle';
import { encodeRecordId } from '../utils/index';

/** The durable `_00_view` row: the last server-confirmed membership. */
export interface DurableView {
  ids: RecordVersionArray;
  confirmed: boolean;
}

export function parseViewRow(row: unknown): DurableView | null {
  if (!row || typeof row !== 'object') return null;
  const ids = (row as { ids?: unknown }).ids;
  if (!Array.isArray(ids)) return null;
  return { ids: ids as RecordVersionArray, confirmed: (row as { confirmed?: unknown }).confirmed === true };
}

/**
 * "Resolved before": the server answered this query on this device at some
 * point. An unconfirmed empty row cannot be told apart from one written before
 * the marker existed, so it does not count.
 */
export const isResolvedBefore = (view: DurableView | null): boolean =>
  view !== null && (view.ids.length > 0 || view.confirmed);

/** The `{ rowCount, state }` object the row-count selects return per row. */
export interface QueryMetaRow {
  id?: RecordId<string>;
  rowCount?: number | null;
  state?: string | null;
}

export function metaFromRow(row: QueryMetaRow | null | undefined): ServerViewMeta {
  if (!row || typeof row !== 'object') return { present: false, rowCount: null, state: null };
  const state = row.state === 'materializing' || row.state === 'ready' ? row.state : null;
  return { present: true, rowCount: typeof row.rowCount === 'number' ? row.rowCount : null, state };
}

/** Collapse duplicate `(id, version)` pairs to one per id, keeping the highest version. */
export function dedupeRecordVersions(pairs: RecordVersionArray): RecordVersionArray {
  if (pairs.length < 2) return pairs;
  const best = new Map<string, number>();
  for (const [id, version] of pairs) {
    const prev = best.get(id);
    if (prev === undefined || version > prev) best.set(id, version);
  }
  if (best.size === pairs.length) return pairs;
  return Array.from(best.entries());
}

export interface MembershipDecisionInput {
  phase: QueryPhase;
  /** Number of ids currently held as membership. */
  held: number;
  remoteArray: RecordVersionArray;
  meta?: ServerViewMeta;
  verifiedRemoval?: boolean;
}

/**
 * Whether a server id-set may be applied. A NON-EMPTY set is always the
 * server's answer. An EMPTY set is believed only when the server stands
 * behind it: the `_00_query` row is present, `ready` (or pre-`state`), and
 * reports zero rows. A missing row while we hold membership is a lost view.
 */
export function decideMembershipOutcome(input: MembershipDecisionInput): MembershipOutcome {
  const { phase, held, remoteArray, meta, verifiedRemoval } = input;
  if (remoteArray.length > 0 || verifiedRemoval) return 'applied';
  if (!meta || !meta.present) {
    return held > 0 || phase !== 'cold' ? 'view-lost' : 'ignored';
  }
  const knownEmpty = meta.rowCount === 0 && meta.state !== 'materializing';
  return knownEmpty ? 'applied' : 'ignored';
}

// ---- list_ref snapshots ------------------------------------------------------

/**
 * A statement that did not answer (an ERR, or no result at all). Kept apart
 * from a `null` result, which is the server saying "no such row": reading a
 * failed statement as a missing row turned every timeout under load into a
 * lost view and re-registered it, over and over (whitepawn, 2026-09-16).
 */
export const FAILED: unique symbol = Symbol('statement failed');
export type Answer<T> = T | null | typeof FAILED;

export interface ListRefEdgeRow {
  in: RecordId<string>;
  out: RecordId<string>;
  version: number;
  parent?: unknown;
}

export interface ListRefSnapshot {
  primary: RecordVersionArray;
  subquery: RecordVersionArray;
  meta: ServerViewMeta;
}

const toPairs = (rows: Array<{ out: RecordId<string>; version: number }> | null | undefined): RecordVersionArray =>
  dedupeRecordVersions(Array.isArray(rows) ? rows.map((r) => [encodeRecordId(r.out), r.version]) : []);

/**
 * Fold the single-query statement batch (edges, meta, children) into a
 * snapshot. `null` when any of the three did not answer: half an answer is
 * not one, and a missing meta would read as a lost view.
 */
export function snapshotFromSingle(
  items: Answer<Array<{ out: RecordId<string>; version: number }>>,
  metaRow: Answer<QueryMetaRow>,
  children: Answer<Array<{ out: RecordId<string>; version: number }>>
): ListRefSnapshot | null {
  if (!Array.isArray(items) || metaRow === FAILED || children === FAILED) return null;
  return { primary: toPairs(items), subquery: toPairs(children), meta: metaFromRow(metaRow) };
}

/**
 * Fold the many-query statement batch into one snapshot per hash. Every hash
 * in `hashById` gets a snapshot; a query whose row did not come back reads
 * `present: false`. Empty when either statement did not answer.
 */
export function snapshotsFromBatch(
  edges: Answer<ListRefEdgeRow[]>,
  counts: Answer<Array<QueryMetaRow | null>>,
  hashById: ReadonlyMap<string, string>
): Map<string, ListRefSnapshot> {
  const out = new Map<string, ListRefSnapshot>();
  if (!Array.isArray(edges) || counts === FAILED) return out;
  for (const hash of hashById.values()) {
    out.set(hash, { primary: [], subquery: [], meta: { present: false, rowCount: null, state: null } });
  }
  for (const row of edges) {
    const hash = hashById.get(encodeRecordId(row.in));
    if (!hash) continue;
    const snapshot = out.get(hash)!;
    const pair: [string, number] = [encodeRecordId(row.out), row.version];
    if (row.parent == null) snapshot.primary.push(pair);
    else snapshot.subquery.push(pair);
  }
  for (const snapshot of out.values()) {
    snapshot.primary = dedupeRecordVersions(snapshot.primary);
    snapshot.subquery = dedupeRecordVersions(snapshot.subquery);
  }
  for (const count of Array.isArray(counts) ? counts : []) {
    if (!count || !count.id) continue;
    const hash = hashById.get(encodeRecordId(count.id));
    if (!hash) continue;
    out.get(hash)!.meta = metaFromRow(count);
  }
  return out;
}

/**
 * A batch that must not be believed at face value: a query we hold rows for
 * whose `_00_query` row did not come back, or no edge at all while the server
 * still reports rows for a held query. Such hashes are re-read one at a time.
 */
export function suspectHashes(
  snapshots: ReadonlyMap<string, ListRefSnapshot>,
  heldCounts: ReadonlyMap<string, number>,
  edgeCount: number
): string[] {
  const out: string[] = [];
  for (const [hash, snapshot] of snapshots) {
    const held = heldCounts.get(hash) ?? 0;
    if (held === 0) continue;
    if (!snapshot.meta.present || (edgeCount === 0 && (snapshot.meta.rowCount ?? 0) > 0)) out.push(hash);
  }
  return out;
}
